tests/test_ac_bytes.py

```diff
@@ -0,0 +1,289 @@
+    ],
+)
+def test_basic_matching(implementation: Optional[Implementation]) -> None:
+    """
+    find_matches_as_indexes() returns matching patterns in the given byte string.
+    """
+    haystack = b"hello, world, hello again"
+    patterns = [b"hello", b"world"]
+    ac = BytesAhoCorasick(patterns, implementation=implementation)
+
+    expected = [b"hello", b"world", b"hello"]
+
+    # find_matches_as_indexes()
+    index_matches = ac.find_matches_as_indexes(haystack)
+    assert [patterns[i] for (i, _, _) in index_matches] == expected
+    assert [haystack[s:e] for (_, s, e) in index_matches] == expected
+
+
+@pytest.mark.parametrize(
+    "implementation",
+    [
+        None,
+        Implementation.NoncontiguousNFA,
+        Implementation.ContiguousNFA,
+        Implementation.DFA,
+    ],
+)
+def test_different_byte_objects_matching(
+    implementation: Optional[Implementation],
+) -> None:
+    """
+    find_matches_as_indexes() returns matching patterns in the given byte string.
+    """
+    haystack = b"hello, world, hello again"
+    patterns = [memoryview(b"hello"), bytearray(b"world")]
+    ac = BytesAhoCorasick(patterns, implementation=implementation)
+
+    expected = [b"hello", b"world", b"hello"]
+
+    # find_matches_as_indexes()
+    index_matches = ac.find_matches_as_indexes(haystack)
+    assert [patterns[i] for (i, _, _) in index_matches] == expected
+    assert [haystack[s:e] for (_, s, e) in index_matches] == expected
+
+
+def test_iterator_of_patterns() -> None:
+    """
+    It's possible to construct ``BytesAhoCorasick()`` with an iterator.
+    """
+    haystack = b"hello, world, hello again"
+    patterns = [b"hello", b"world"]
+    ac = BytesAhoCorasick(iter(patterns))
+
+    expected = [b"hello", b"world", b"hello"]
+
+    index_matches = ac.find_matches_as_indexes(haystack)
+    assert [patterns[i] for (i, _, _) in index_matches] == expected
+    assert [haystack[s:e] for (_, s, e) in index_matches] == expected
+
+
+def test_bad_iterators() -> None:
+    """
+    When constructed with a bad iterator, the underlying Python error is raised.
+    """
+    with pytest.raises(TypeError):
+        BytesAhoCorasick(None)  # type: ignore
+
+    with pytest.raises(TypeError):
+        BytesAhoCorasick([b"x", 12])  # type: ignore[list-item]
+
+    # str doesn't implement the buffer API and can't be converted to bytes
+    with pytest.raises(TypeError):
+        BytesAhoCorasick([b"x", "y"])  # type: ignore[list-item]
+
+
+@given(
+    st.lists(st.binary(min_size=3), min_size=1, max_size=30_000),
+)
+def test_construction_extensive(patterns: list[bytes]) -> None:
+    """
+    Exercise the construction code paths, ensuring we end up using all
+    patterns.
+    """
+    patterns = [b"@%b@" % p for p in patterns]
+    ac = BytesAhoCorasick(patterns)
+    for i, p in enumerate(patterns):
+        assert [p[s:e] for (_, s, e) in ac.find_matches_as_indexes(p)] == [p]
+
+
+@pytest.mark.parametrize(
```

Comment:
Can probably delete this test, the unitary nature of unicode characters is irrelevant here.